### PR TITLE
Enhanced speaker counting for short audio recordings

### PR DESCRIPTION
**Pull request by tango4j**
This pull request is for adding the function getEnhancedSpeakerCount() that performs an enhanced speaker counting for speaker diarization module.

It improves speaker counting accuracy from 50% to 80% especially for short (less than 1 min) recordings.

### CODE REVIEW

File: nemo/collections/asr/parts/utils/nmse_clustering.py [hunk 1]
```diff
@@ -426,7 +482,16 @@ def getPvalueList(self):
...
 
 
-def COSclustering(key, emb, oracle_num_speakers=None, max_num_speaker=8, min_samples=6, fixed_thres=None, cuda=False):
+def COSclustering(
+    key,
+    emb,
+    oracle_num_speakers=None,
+    max_num_speaker=8,
+    min_samples=6,
+    enhanced_count_thres=80,
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmse_clustering.py:491]: Added doc string for enhanced_count_thres

File: nemo/collections/asr/parts/utils/nmse_clustering.py [hunk 2]
```diff
@@ -445,16 +510,21 @@ def COSclustering(key, emb, oracle_num_speakers=None, max_num_speaker=8, min_sam
...
+            it is directed to the enhanced speaker counting mode.
     Returns:
         Y: (List[int])
             Speaker label for each segment.
     """
+
+    if emb.shape[0] == 1:
+        return np.array([0])
+    elif emb.shape[0] < enhanced_count_thres and oracle_num_speakers == None:
+        oracle_num_speakers = getEnhancedSpeakerCount(key, emb, cuda)
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmse_clustering.py:523]: Introduced a new variable est_num_of_spk_enhanced to make it clear.
If oracle_num_speakers and est_num_of_speak both doesn't exist, we use est_num_of_spk.

File: nemo/collections/asr/parts/utils/nmse_clustering.py [hunk 3]
```diff
@@ -118,6 +121,63 @@ def getMinimumConnection(mat, max_N, n_list):
...
+    return new_emb_np, anchor_sample_n * anchor_spk_n
+
+
+def getEnhancedSpeakerCount(key, emb, cuda, random_test_count=5, anchor_spk_n=3, anchor_sample_n=10, sigma=50):
+    """
+    Calculates the number of speakers using NME analysis with anchor embeddings.
+    """
+    est_num_of_spk_list = []
+    for seed in range(random_test_count):
+        np.random.seed(seed)
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmse_clustering.py:160]: Changing np to torch is on hold.

File: nemo/collections/asr/parts/utils/nmse_clustering.py [hunk 4]
```diff
@@ -118,6 +121,63 @@ def getMinimumConnection(mat, max_N, n_list):
...
+        The number of speakers for synthetic embedding
+
+    sigma (int):
+        The amplitude of synthetic noise for each embedding vector
+
+    """
+    emb_dim = emb.shape[1]
+    mean, std_org = np.mean(emb, axis=0), np.std(emb, axis=0)
+    new_emb_list = []
+    for _ in range(anchor_spk_n):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmse_clustering.py:142]: Changing np to torch is on hold.

File: nemo/collections/asr/parts/utils/nmse_clustering.py [hunk 5]
```diff
@@ -118,6 +121,63 @@ def getMinimumConnection(mat, max_N, n_list):
     return affinity_mat, p_value
 
 
+def addAnchorEmb(emb, anchor_sample_n, anchor_spk_n, sigma):
+    """
+    Add randomly generated synthetic embeddings to make eigen analysis more stable.
+    We refer to these embeddings as anchor embeddings.
+
+    anchor_sample_n (int):
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/parts/utils/nmse_clustering.py:129]: Added doc string for emb and also added recommended values for the arguments.

**Review by nithinraok — APPROVED** [review]
Looks good to me, thanks

**Pull request merged**